### PR TITLE
firefox-{beta,devedition}-unwrapped: init at 112.0b6

### PR DESCRIPTION
**Pull request by jopejoe1**
###### Description of changes
Added Firefox beta and devedition to be build from source, this makes it so that they can also be used on architectures that do not have a bin version like aarch64-linux.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 1]
```diff
@@ -389,7 +390,7 @@ buildStdenv.mkDerivation ({
     (enableFeature enableDebugSymbols "debug-symbols")
   ]
   ++ lib.optionals enableDebugSymbols [ "--disable-strip" "--disable-install-strip" ]
-  ++ lib.optional enableOfficialBranding "--enable-official-branding"
+  ++ lib.optionals enableOfficialBranding [ "--enable-official-branding" "--with-branding=browser/branding/${branding}" ]
```
Comment (mweinelt) [inline pkgs/applications/networking/browsers/firefox/common.nix:393]: Ideally we could stick with the default branding value for most consumers of `buildMozillaMach`.

Please note that this function serves firefox, firefox-esr, librewolf, thunderbird, soon apparently also betterbird, and at least the mail clients are left out here. 

If we make the branding overridable, then the whole path (`browser/branding/official`) needs to be configurable, not just its final segment.

For most causes not setting the branding does work, so I would only want it to be set in cases where it does not work out.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 2]
```diff
@@ -5,6 +5,7 @@
 , binaryName ? "firefox"
 , application ? "browser"
 , applicationName ? "Mozilla Firefox"
+, branding ? "official"
```
Comment (mweinelt) [inline pkgs/applications/networking/browsers/firefox/common.nix:8]:
```suggestion
, branding ? null
```

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 1 above)
Comment (mweinelt) [inline pkgs/applications/networking/browsers/firefox/common.nix:393]:
```suggestion
  ++ lib.optional enableOfficialBranding "--enable-official-branding"
  ++ lib.optional (branding != null) "--with-branding=${branding}"
```

File: pkgs/applications/networking/browsers/firefox/packages.nix [hunk 3]
```diff
@@ -27,6 +27,62 @@ rec {
...
+      attrPath = "firefox-beta-unwrapped";
+      versionSuffix = "b[0-9]*";
+    };
+  };
+
+  firefox-devedition = buildMozillaMach rec {
+    pname = "firefox-devedition";
+    version = "112.0b5";
+    applicationName = "Mozilla Firefox Developer Edition";
+    branding = "aurora";
```
Comment (mweinelt) [inline pkgs/applications/networking/browsers/firefox/packages.nix:61]:
```suggestion
    branding = "browser/branding/aurora";
```

**Review by mweinelt — CHANGES_REQUESTED** [review]
I think if we do it like this, it doesn't affect builds, that are fine with their default branding.

**Comment by mweinelt** [comment]
Do you intend to maintain these packages?

**Comment by jopejoe1** [comment]
> I think if we do it like this, it doesn't affect builds, that are fine with their default branding.

That would work, applied it to the pr.

> Do you intend to maintain these packages?

Yes, I do intend to maintain them.

**Comment by K900** [comment]
Do we really need both beta and dev edition? The changes between the two are very minimal, and the build time effectively doubles.

**Comment by mweinelt** [comment]
And changes to `firefox/common.nix` should go into a separate commit that is prefixed with `buildMozillaMach:` and a proper description of the change and its requirement.

**Comment by mweinelt** [comment]
> Do we really need both beta and dev edition? The changes between the two are very minimal, and the build time effectively doubles.

The effective problem here is, that we expect to build the beta/devedition more often, than we have users consuming it.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 4]
```diff
@@ -5,6 +5,8 @@
 , binaryName ? "firefox"
 , application ? "browser"
 , applicationName ? "Mozilla Firefox"
+# Optional path of which branding should be used
```
[the remark this answers is not in the document]
Comment (jopejoe1) [reply pkgs/applications/networking/browsers/firefox/common.nix]: Removed it.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 222701` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>firefox-beta-unwrapped (firefoxPackages.firefox-beta)</li>
    <li>firefox-beta-unwrapped.debug (firefoxPackages.firefox-beta.debug)</li>
    <li>firefox-beta-unwrapped.symbols (firefoxPackages.firefox-beta.symbols)</li>
    <li>firefox-devedition-unwrapped (firefoxPackages.firefox-devedition)</li>
    <li>firefox-devedition-unwrapped.debug (firefoxPackages.firefox-devedition.debug)</li>
    <li>firefox-devedition-unwrapped.symbols (firefoxPackages.firefox-devedition.symbols)</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
Tested both packages on aarch64-linux and x86_64-linux.

**Comment by lucasew** [comment]
The derivations are working. Tested installing uBlock Origin and checking if it works in some sites.

It worked nice on my i5 7200U laptop

**Review by lucasew — APPROVED** [review]

**Comment by mweinelt** [comment]
The attribute name need to be `firefox-beta-unwrapped` and `firefox-devedition-unwrapped`, because that is where the passthru tests are also attached. Please update the commits and keep that in mind.

Also, 112.0b6 is out.

**Comment by jopejoe1** [comment]
> The attribute name need to be `firefox-beta-unwrapped` and `firefox-devedition-unwrapped`, because that is where the passthru tests are also attached. Please update the commits and keep that in mind.
> 
> Also, 112.0b6 is out.

Changed commit messages to include `-unwrapped` and updated the version to `112.0b6`.

**Pull request merged**